src/StreamJsonRpc/WebSocketMessageHandler.cs

```diff
@@ -0,0 +1,137 @@
+                char[] decodedChars = new char[this.readBuffer.Array.Length];
+                void DecodeInput()
+                {
+                    int decodedCharsCount = this.readDecoder.GetChars(this.readBuffer.Array, 0, result.Count, decodedChars, 0, result.EndOfMessage);
+                    jsonBuilder.Append(decodedChars, 0, decodedCharsCount);
+                }
+
+                DecodeInput();
+                while (!result.EndOfMessage)
+                {
+                    result = await this.WebSocket.ReceiveAsync(this.readBuffer, cancellationToken).ConfigureAwait(false);
+                    DecodeInput();
+                }
+
+                return jsonBuilder.ToString();
+            }
+        }
+
+        /// <inheritdoc />
+        protected async override Task WriteCoreAsync(string content, Encoding contentEncoding, CancellationToken cancellationToken)
+        {
+            Requires.NotNull(content, nameof(content));
+            Requires.NotNull(contentEncoding, nameof(contentEncoding));
+
+            if (contentEncoding.GetByteCount(content) <= this.writeBuffer.Length)
+            {
+                // Fast path: send the whole message as a single chunk.
+                int bytesWritten = contentEncoding.GetBytes(content, 0, content.Length, this.writeBuffer, 0);
+                var bufferSegment = new ArraySegment<byte>(this.writeBuffer, 0, bytesWritten);
+                await this.WebSocket.SendAsync(bufferSegment, WebSocketMessageType.Text, true, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                var encoder = contentEncoding.GetEncoder();
+                bool completed;
+                int bytesUsed;
+                int charsLeftToConvert = content.Length;
+                while (charsLeftToConvert > 0)
+                {
+                    unsafe
+                    {
+                        fixed (byte* writeBuffer = this.writeBuffer)
+                        fixed (char* pContent = content)
+                        {
+                            char* pStart = pContent + content.Length - charsLeftToConvert;
+                            encoder.Convert(pStart, charsLeftToConvert, writeBuffer, this.writeBuffer.Length, false, out int charsUsed, out bytesUsed, out completed);
+                            charsLeftToConvert -= charsUsed;
+                        }
+                    }
+
+                    var bufferSegment = new ArraySegment<byte>(this.writeBuffer, 0, bytesUsed);
+                    await this.WebSocket.SendAsync(bufferSegment, WebSocketMessageType.Text, completed, cancellationToken).ConfigureAwait(false);
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                this.WebSocket.Dispose();
+            }
+
+            base.Dispose(disposing);
+        }
```

Comment:
Hmmm...I think this is a good idea...

Reply:
Do you mean you support disposing the `WebSocket` by default? What do you think of the issue with needing an async handshake to disconnect though, per my first comment on the PR to you?

Reply:
I really am not expert on WebSocket semantics.  Do you have someone you can consult with more experience?

---
In reply to: [160271489](https://github.com/Microsoft/vs-streamjsonrpc/pull/81#discussion_r160271489) [](ancestors = 160271489)